plugins/kafka-records-storage/src/main/java/com/github/bsideup/liiklus/kafka/KafkaRecordsStorage.java

```diff
@@ -85,69 +99,141 @@ public Subscription subscribe(String topic, String groupName, Optional<String> a
+                            new ByteBufferDeserializer(),
+                            new ByteBufferDeserializer()
+                    );
+
+                    val topics = Arrays.asList(topic);
+
+                    if (!sink.isCancelled()) {
+                        val pausedPartitions = new ConcurrentHashMap<TopicPartition, Boolean>();
+
+                        val records = Flux
+                                .<ConsumerRecords<ByteBuffer, ByteBuffer>>create(recordsSink -> {
+                                    try {
+                                        while (!sink.isCancelled() && !recordsSink.isCancelled()) {
+                                            for (val entry : pausedPartitions.entrySet()) {
+                                                val topicPartition = entry.getKey();
+                                                if (!consumer.assignment().contains(topicPartition)) {
+                                                    continue;
+                                                }
+
+                                                val partitions = Arrays.asList(topicPartition);
+                                                if (entry.getValue()) {
+                                                    consumer.pause(partitions);
+                                                } else {
+                                                    consumer.resume(partitions);
+                                                }
+                                            }
+
+                                            val consumerRecords = consumer.poll(Duration.ofMillis(10));
+                                            if (!consumerRecords.isEmpty()) {
+                                                recordsSink.next(consumerRecords);
+                                            }
+                                        }
+
+                                        try {
+                                            consumer.unsubscribe();
+                                        } catch (Exception e) {
+                                            log.warn("Failed to unsubscribe", e);
+                                        }
+                                        try {
+                                            consumer.close();
+                                        } catch (Exception e) {
+                                            log.warn("Failed to close", e);
+                                        }
+
+                                        recordsSink.complete();
+                                    } catch (Exception e) {
+                                        recordsSink.error(e);
+                                    }
+                                }, FluxSink.OverflowStrategy.ERROR)
+                                .subscribeOn(KAFKA_POLL_SCHEDULER)
+                                .publish(1);
+
+                        val revocations = new HashMap<Integer, DirectProcessor<Boolean>>();
```

Comment:
should it be ConcurrentHashMap? as we update it asynchronously

Reply:
There is no concurrent write access to `revocations` since `ConsumerRebalanceListener` calls only one `onPartitions*` at a time